gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/strategy/finalization/DetachStrategy.java

```diff
@@ -0,0 +1,186 @@
+import org.apache.tinkerpop.gremlin.process.traversal.step.sideEffect.ProfileSideEffectStep;
+import org.apache.tinkerpop.gremlin.process.traversal.step.util.EmptyStep;
+import org.apache.tinkerpop.gremlin.process.traversal.strategy.AbstractTraversalStrategy;
+import org.apache.tinkerpop.gremlin.process.traversal.util.TraversalHelper;
+import org.apache.tinkerpop.gremlin.structure.util.detached.DetachedFactory;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+/**
+ * A strategy that manages the properties that will be in the result.
+ *
+ * @author Valentyn Kahamlyk
+ */
+public final class DetachStrategy extends AbstractTraversalStrategy<TraversalStrategy.FinalizationStrategy> implements TraversalStrategy.FinalizationStrategy {
+
+    private static final DetachStrategy INSTANCE = new DetachStrategy(DetachMode.ALL, null);
+    private DetachMode detachMode = DetachMode.ALL;
+    private List<String> properties;
+
+    private DetachStrategy() {}
+
+    private DetachStrategy(final DetachMode detachMode, final List<String> properties) {
+        this.detachMode = detachMode;
+        this.properties = properties;
+    }
+
+    private DetachStrategy(final Builder builder) {
+        this(builder.detachMode, builder.properties);
+    }
+
+    @Override
+    public void apply(final Traversal.Admin<?, ?> traversal) {
+        if (traversal.getParent() == EmptyStep.instance()) {
+            final Optional<ProfileSideEffectStep> profileStep = TraversalHelper.getFirstStepOfAssignableClass(ProfileSideEffectStep.class, traversal);
+            final int index = profileStep.map(step -> traversal.getSteps().indexOf(step))
+                    .orElseGet(() -> traversal.getSteps().size());
+            traversal.addStep(index,
+                    new DetachElementStep<>(traversal, new DetachOptions(detachMode, properties)));
+        }
+    }
+
+    public static DetachStrategy create(final Configuration configuration) {
+        return new DetachStrategy(DetachMode.valueOf(configuration.getString(ID_MODE)),
+                new ArrayList<>((Collection<String>) configuration.getProperty(ID_PROPERTIES)));
+    }
+
+    /**
+     * Gets the standard configuration of this strategy that will return all properties.
+     */
+    public static DetachStrategy instance() {
+        return INSTANCE;
+    }
+
+    public static final String ID_MODE = "detachMode";
+    public static final String ID_PROPERTIES = "properties";
+
+    @Override
+    public Configuration getConfiguration() {
+        final Map<String, Object> map = new HashMap<>();
+        map.put(STRATEGY, DetachStrategy.class.getCanonicalName());
+        map.put(ID_MODE, this.detachMode.toString());
+        map.put(ID_PROPERTIES, this.properties);
+        return new MapConfiguration(map);
+    }
+
+    public enum DetachMode {
```

Comment:
it's too bad this enum addition forces all those serializer changes. is there other precedent for this in other strategies? in other words, do any other strategies introduce their own enums? maybe we can avoid this with some shift in the way we do the syntax?

Reply:
There are no other strategies that would add enums. May I just use a string value?

Reply:
we do have a precedent for using integer constants for `with()`-step: https://github.com/apache/tinkerpop/blob/master/gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/step/util/WithOptions.java 

maybe we should follow that pattern here and save the serialization expansion. maybe the rule is to save enum for core language elements, like `Scope`, `Direction`, etc.